impl/monitor/monitor.go

```diff
@@ -0,0 +1,247 @@
+		glog.Errorf("s.pageMutations(_): %v", err)
+		return nil, err
+	}
+
+	// update seen SMRs:
+	// TODO: this should be
+
+	rh := resp.GetSmr().GetRootHash()
+	switch err := s.verifyMutations(mutations, rh); err {
+	// TODO(ismail): return proper data for failure cases:
+	case ErrInvalidMutation:
+		glog.Errorf("TODO: handle this ErrInvalidMutation properly")
+	case ErrInvalidSignature:
+		glog.Errorf("TODO: handle this ErrInvalidSignature properly (return SMR)")
+	case ErrNotMatchingRoot:
+		glog.Errorf("TODO: handle this ErrNotMatchingRoot properly")
+	case nil:
+		glog.Info("TODO: Successfully verified.")
+	default:
+		glog.Errorf("Unexpected error: %v", err)
+	}
+
+	// TODO(ismail): Sign reconstructed Hash instead of response hash:
+	sig, err := s.signer.SignObject(resp.GetSmr().GetRootHash())
+	if err != nil {
+		return nil, fmt.Errorf("s.signer.SignObject(_): %v", err)
+	}
+	// Update seen/processed signed map roots:
+	s.proccessedSMRs = append(s.proccessedSMRs,
+		&ktpb.GetMonitoringResponse{
+			Smr: &trillian.SignedMapRoot{
+				Signature:      sig,
+				TimestampNanos: resp.GetSmr().GetTimestampNanos(),
+				RootHash:       resp.GetSmr().GetRootHash(),
+				Metadata:       resp.GetSmr().GetMetadata(),
+				MapId:          resp.GetSmr().GetMapId(),
+				MapRevision:    resp.GetSmr().GetMapRevision(),
+			},
+			IsValid:            true,
+			SeenTimestampNanos: seen,
+		})
+
+	return mutations, nil
+}
+
+func (s *Server) verifyMutations(ms []*ktpb.Mutation, expectedRoot []byte) error {
+	// TODO(ismail):
+	// For each received mutation in epoch e:
+	// Verify that the provided leaf’s inclusion proof goes to epoch e -1.
+	// Verify the mutation’s validity against the previous leaf.
+	// Compute the new leaf and store the intermediate hashes locally.
+	// Compute the new root using local intermediate hashes from epoch e.
+	for _, m := range ms {
+		idx := m.GetProof().GetLeaf().GetIndex()
+		nbrs := m.GetProof().GetInclusion()
+		if err := s.tree.VerifyProof(nbrs, idx, m.GetProof().GetLeaf().GetLeafValue(),
+			sparse.FromBytes(expectedRoot)); err != nil {
+			glog.Errorf("VerifyProof(): %v", err)
+			// TODO return nil, err
+		}
+	}
+
+	return nil
+}
+
+// pageMutations iterates/pages through all mutations in the case there were
+// more then maximum pageSize mutations in between epochs.
+// It will modify the passed GetMutationsResponse resp and the passed list of
+// mutations ms.
+func (s *Server) pageMutations(ctx context.Context, resp *ktpb.GetMutationsResponse,
+	ms []*ktpb.Mutation, opts ...grpc.CallOption) error {
+	// Query all mutations in the current epoch
+	for resp.GetNextPageToken() != "" {
+		req := &ktpb.GetMutationsRequest{PageSize: pageSize}
+		resp, err := s.client.GetMutations(ctx, req, opts...)
+		if err != nil {
+			return err
+		}
+		ms = append(ms, resp.GetMutations()...)
+	}
+	return nil
+}
+
+func (s *Server) lastSeenSMR() *trillian.SignedMapRoot {
```

Comment:
Functions that implement the proto interface also probably deserve their own file